tracking/reporter.cpp

```diff
@@ -0,0 +1,147 @@
+#include "reporter.hpp"
+
+#include "base/logging.hpp"
+#include "base/thread.hpp"
+#include "base/timer.hpp"
+
+#include "boost/circular_buffer.hpp"
+
+#include "platform/location.hpp"
+#include "platform/socket.hpp"
+
+#include "std/mutex.hpp"
+#include "std/vector.hpp"
+
+#include "tracking/connection.hpp"
+
+using namespace tracking;
+
+namespace
+{
+double constexpr kMinDelaySeconds = 1.0;
+double constexpr kReconnectDelaySeconds = 60.0;
+size_t constexpr kRealTimeBufferSize = 60;
+
+class WorkerImpl final : public Reporter::Worker
+{
+public:
+  WorkerImpl(unique_ptr<platform::Socket> socket, size_t pushDelayMs);
+  void Run();
+
+  // Worker overrides
+  void AddLocation(location::GpsInfo const & info);
+  void Stop();
+
+private:
+  void FetchInput();
+  bool SendPoints();
+
+  volatile bool m_stop = false;
```

Comment:
If you need some kind of cancellation event, we have base/cancellable.hpp for such purposes. Don't use volatile for synchronization.